### PR TITLE
Fcrepo 2708/2709 - Binary and Description Mementos

### PR DESCRIPTION
**Pull request by bbpennel**
**JIRA Tickets**: 
https://jira.duraspace.org/browse/FCREPO-2708
https://jira.duraspace.org/browse/FCREPO-2709

# What does this Pull Request do?
Adds support for creating binary and descriptions mementos from request body.

# What's new?
* Binary mementos can be created from submitted body or existing resource.
* Description mementos can be created from existing resource and submitted bodies. Can be created without a binary memento.
* When a binary memento is created from original resource, description is also created
* Binary mementos use content type from description memento at the same datetime if available, otherwise the default mimetype is returned.
* No longer uses modeshape copy to create binary mementos from existing resources
* describes and describedby links for mementos reference originals
* Binary properties provided with memento creation request are ignored since they will be supplied by the submitted description.
* Added helper method for FedoraResource to getOriginalResource
* Refactor to consolidate binary and description creation in binary service rather than duplicating in version service.

# How should this be tested?
```
# Create versioned binary
curl -i -XPOST -H "Slug:versioned_bin11" -H "Link: <http://mementoweb.org/ns#OriginalResource>; rel=\"type\"" -H "Content-Type: text/plain" --data-binary "binarycontent" "http://localhost:8080/"

# Create historic memento with content
curl -XPOST http://localhost:8080/versioned_bin11/fcr:versions -H "Memento-Datetime: Sat, 1 Jan 2000 00:00:00 GMT" -H "Content-Type: text/plain" --data-binary "text content"

# Create historic memento with no content
curl -XPOST http://localhost:8080/versioned_bin11/fcr:versions -H "Memento-Datetime: Sat, 1 Jan 2000 01:00:00 GMT"

# Create memento from current version
curl -XPOST http://localhost:8080/versioned_bin11/fcr:versions

curl -v <uri from response of last command>
# Memento-Datetime: <should be roughly current time>
# Content-Type: text/plain
# Describedby: http://localhost:8080/versioned_bin11/fcr:metadata
# "binarycontent"

curl -v <uri from response of last command, insert "/fcr:metadata" before "/fcr:versions">
# Should return normal RDF response with mimetype = text/plain

curl -v http://localhost:8080/versioned_bin11/fcr:versions/20000101000000
# Memento-Datetime: Sat, 1 Jan 2000 00:00:00 GMT
# Content-Type: application/octet-stream (content type provided when making memento is ignored)
# "text content"

curl -v http://localhost:8080/versioned_bin11/fcr:versions/20000101010000
# Memento-Datetime: Sat, 1 Jan 2000 01:00:00 GMT
# Content-Type: application/octet-stream
# ""

# Check that the original is OK
curl -v http://localhost:8080/versioned_bin11
# Content-Type: text/plain
# "binarycontent"



curl -v http://localhost:8080/versioned_bin11/fcr:metadata/fcr:versions/20000101000000
# Status: 404 (not created yet)

echo '<http://localhost:8080/versioned_bin11> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://fedora.info/definitions/v4/repository#NonRdfSourceDescription> .
<http://localhost:8080/versioned_bin11> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://fedora.info/definitions/v4/repository#Binary> .
<http://localhost:8080/versioned_bin11> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://fedora.info/definitions/v4/repository#Resource> .
<http://localhost:8080/versioned_bin11> <http://fedora.info/definitions/v4/repository#lastModifiedBy> "bypassAdmin" .
<http://localhost:8080/versioned_bin11> <http://www.loc.gov/premis/rdf/v1#hasSize> "13"^^<http://www.w3.org/2001/XMLSchema#long> .
<http://localhost:8080/versioned_bin11> <http://www.ebu.ch/metadata/ontologies/ebucore/ebucore#hasMimeType> "text/awesome" .
<http://localhost:8080/versioned_bin11> <http://fedora.info/definitions/v4/repository#createdBy> "bypassAdmin" .
<http://localhost:8080/versioned_bin11> <http://fedora.info/definitions/v4/repository#created> "2018-04-26T19:43:12.137Z"^^<http://www.w3.org/2001/XMLSchema#dateTime> .
<http://localhost:8080/versioned_bin11> <http://www.loc.gov/premis/rdf/v1#hasMessageDigest> <urn:sha1:23dec4c4f917668fdbcc8ea717b0b6f5a3395d51> .
<http://localhost:8080/versioned_bin11> <http://fedora.info/definitions/v4/repository#lastModified> "2018-04-26T19:43:12.137Z"^^<http://www.w3.org/2001/XMLSchema#dateTime> .
<http://localhost:8080/versioned_bin11> <http://www.ebu.ch/metadata/ontologies/ebucore/ebucore#filename> "" .
<http://localhost:8080/versioned_bin11> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/ns/ldp#NonRDFSource> .
<http://localhost:8080/versioned_bin11> <http://fedora.info/definitions/v4/repository#writable> "true"^^<http://www.w3.org/2001/XMLSchema#boolean> .
<http://localhost:8080/versioned_bin11> <http://www.iana.org/assignments/relation/describes> <http://localhost:8080/versioned_bin11> .
<http://localhost:8080/versioned_bin11> <http://fedora.info/definitions/v4/repository#hasFixityService> </fcr:fixity> .' | curl -XPOST -H "Memento-Datetime: Sat, 1 Jan 2000 00:00:00 GMT" http://localhost:8080/versioned_bin11/fcr:metadata/fcr:versions -H "Content-Type: application/n-triples" --data-binary "@-"

curl -v http://localhost:8080/versioned_bin11/fcr:metadata/fcr:versions/20000101000000
# Should return subjected RDF with mimetype = text/awesome

curl -v http://localhost:8080/versioned_bin11/fcr:versions/20000101000000
# Memento-Datetime: Sat, 1 Jan 2000 00:00:00 GMT
# Content-Type: text/awesome (content type should now be coming from description)
# "text content"

# Create description memento from original
curl -XPOST http://localhost:8080/versioned_bin11/fcr:metadata/fcr:versions

# Verify description memento created
curl -v <uri from response of last command>
# rdf matching original, mimetype = text/plain

curl -v <uri from response of previous command with "/fcr:metadata" removed>
# status: 404

```

# Additional Notes:
I intend to see about turning on more of the disabled tests in FedoraVersioningIT, but wanted to get PR created.

# Interested parties
@fcrepo4/committers

### CODE REVIEW

File: fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/FedoraBinaryImpl.java [hunk 1]
```diff
@@ -100,18 +100,38 @@ public FedoraBinaryImpl(final Node node) {
...
         return new NonRdfSourceDescriptionImpl(getDescriptionNode());
     }
 
     @Override
     protected Node getDescriptionNode() {
         try {
+            final Node node = getNode();
+            if (isMemento()) {
+                final String mementoName = node.getName();
+                return node.getNode("../../" + FEDORA_DESCRIPTION + "/" + LDPCV_TIME_MAP + "/" + mementoName);
```
Comment (whikloj) [inline fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/FedoraBinaryImpl.java]: This makes me cringe, also isn't the Binary just one `../` in front of the `FEDORA_DESCRIPTION`?

File: fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/FedoraBinaryImpl.java [hunk 2]
```diff
@@ -100,18 +100,38 @@ public FedoraBinaryImpl(final Node node) {
...
+            }
             return getNode().getNode(FEDORA_DESCRIPTION);
         } catch (final RepositoryException e) {
             throw new RepositoryRuntimeException(e);
         }
     }
 
+    private Node getDescriptionNodeOrNull() {
```
Comment (whikloj) [inline fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/FedoraBinaryImpl.java]: Why would there not be a description node? It is created as part of the Binary, right?

File: fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/FedoraBinaryImpl.java (hunk 1 above)
Comment (bbpennel) [reply fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/FedoraBinaryImpl.java]: Yeah its pretty cringy. I started writing it with a pile of `getParent`s and `getNode`s, but it didn't feel more readable to me.  I'm fine with changing it if others prefer though.

File: fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/FedoraBinaryImpl.java (hunk 2 above)
Comment (bbpennel) [reply fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/FedoraBinaryImpl.java]: For the annoying period between creating the historic binary memento and importing the description for it. Rather than creating a description and allowing the user to update it, this PR simply doesn't create it until the client provides it.

For mementos created from the original binary resource, the description is automatically created.

File: fcrepo-http-api/src/main/java/org/fcrepo/http/api/FedoraVersioning.java [hunk 3]
```diff
@@ -185,29 +186,40 @@ public Response addVersion(@HeaderParam(MEMENTO_DATETIME_HEADER) final String da
...
                 // Create memento
-                final FedoraResource memento;
-                if (resource instanceof FedoraBinary) {
-                    memento = versionService.createBinaryVersion(session.getFedoraSession(), resource(),
-                            mementoInstant, null, null, null, null);
-                } else {
+                FedoraResource memento = null;
+                final boolean isBinary = resource instanceof FedoraBinary;
+                if (isBinary) {
+                    final FedoraBinary binaryResource = (FedoraBinary) resource;
```
Comment (whikloj) [inline fcrepo-http-api/src/main/java/org/fcrepo/http/api/FedoraVersioning.java]: This is probably my lack of understanding, but if `isBinary == true` because `resource instanceof FedoraBinary` then why this line?

Comment (bbpennel) [reply fcrepo-http-api/src/main/java/org/fcrepo/http/api/FedoraVersioning.java]: need to pass a FedoraBinary object into `versionService.createBinaryVersion`, so I cast it once for passing around. Unless I'm misunderstanding the question.

File: fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/services/BinaryService.java [hunk 4]
```diff
@@ -17,12 +17,32 @@
...
+    FedoraBinary findOrCreateBinary(FedoraSession session, String path);
+
+    /**
+     * Retrieves a binary description instance by session and path.
+     *
+     * @param session session
+     * @param path path of description
+     * @return retrieved NonRdfSourceDescription
+     */
+    NonRdfSourceDescription findOrCreateDescription(FedoraSession session, String path);
```
[the remark this answers is not in the document]
Comment (bbpennel) [reply fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/services/BinaryService.java]: In this branch a fedora:description can exist without being the child of a binary, as is the case for description mementos (they are children of `fedora:description/fedora:timemap`).  I'm guessing that calling findOrCreateBinary on the timemap might do bad things

**Comment by whikloj** [comment]
Can't delete a binary or binary description Memento and then re-use the same date. Gets 409 Conflict.
```
> curl -i -XGET -ufedoraAdmin:fedoraAdmin '-HAccept: application/link-format' http://localhost:8080/rest/test_versioning/0e/4b/a0/88/0e4ba088-ea77-4760-99ab-d57ead90bb6e/fcr:metadata/fcr:versions
HTTP/1.1 200 OK
Date: Fri, 27 Apr 2018 20:42:30 GMT
Set-Cookie: JSESSIONID=g9qz2ajembld6ri2kdpha07d;Path=/
Expires: Thu, 01 Jan 1970 00:00:00 GMT
Set-Cookie: rememberMe=deleteMe; Path=/; Max-Age=0; Expires=Thu, 26-Apr-2018 20:42:30 GMT
ETag: W/"6b0a366757784c1859f29d2755a6b58c2e4008c7"
Last-Modified: Fri, 27 Apr 2018 20:42:30 GMT
Link: <http://www.w3.org/ns/ldp#Resource>;rel="type"
Link: <http://www.w3.org/ns/ldp#RDFSource>;rel="type"
Link: <http://localhost:8080/rest/test_versioning/0e/4b/a0/88/0e4ba088-ea77-4760-99ab-d57ead90bb6e/fcr:metadata/fcr:versions/fcr:acl>; rel="acl"
Link: <http://mementoweb.org/ns#TimeMap>; rel="type"
Link: <http://localhost:8080/rest/test_versioning/0e/4b/a0/88/0e4ba088-ea77-4760-99ab-d57ead90bb6e/fcr:metadata>; rel="original"
Link: <http://localhost:8080/rest/test_versioning/0e/4b/a0/88/0e4ba088-ea77-4760-99ab-d57ead90bb6e/fcr:metadata>; rel="timegate"
Link: <http://localhost:8080/rest/test_versioning/0e/4b/a0/88/0e4ba088-ea77-4760-99ab-d57ead90bb6e/fcr:metadata/fcr:versions>; rel="timemap"
Vary-Post: Memento-Datetime
Allow: POST,HEAD,GET,OPTIONS,DELETE
Content-Type: application/link-format
Content-Length: 850
Server: Jetty(9.3.1.v20150714)

<http://localhost:8080/rest/test_versioning/0e/4b/a0/88/0e4ba088-ea77-4760-99ab-d57ead90bb6e/fcr:metadata>; rel="original",
<http://localhost:8080/rest/test_versioning/0e/4b/a0/88/0e4ba088-ea77-4760-99ab-d57ead90bb6e/fcr:metadata>; rel="timegate",
<http://localhost:8080/rest/test_versioning/0e/4b/a0/88/0e4ba088-ea77-4760-99ab-d57ead90bb6e/fcr:versions/20180427204227>; rel="memento"; datetime="Fri, 27 Apr 2018 20:42:27 GMT",
<http://localhost:8080/rest/test_versioning/0e/4b/a0/88/0e4ba088-ea77-4760-99ab-d57ead90bb6e/fcr:versions/20000523173445>; rel="memento"; datetime="Tue, 23 May 2000 17:34:45 GMT",
<http://localhost:8080/rest/test_versioning/0e/4b/a0/88/0e4ba088-ea77-4760-99ab-d57ead90bb6e/fcr:metadata/fcr:versions>; rel="self"; from="Tue, 23 May 2000 17:34:45 GMT"; until="Fri, 27 Apr 2018 20:42:27 GMT"; type="application/link-format",

[/sw/var/www/fcrepo4/fedora4-tests] 
curl -s -o /dev/null --write-out '%{http_code}' --no-keepalive -i -XDELETE -ufedoraAdmin:fedoraAdmin http://localhost:8080/rest/test_versioning/0e/4b/a0/88/0e4ba088-ea77-4760-99ab-d57ead90bb6e/fcr:versions/20000523173445
204

[/sw/var/www/fcrepo4/fedora4-tests] 
> curl -ufedoraAdmin:fedoraAdmin -H"Accept: application/link-format" http://localhost:8080/rest/test_versioning/0e/4b/a0/88/0e4ba088-ea77-4760-99ab-d57ead90bb6e/fcr:versions
<http://localhost:8080/rest/test_versioning/0e/4b/a0/88/0e4ba088-ea77-4760-99ab-d57ead90bb6e>; rel="original",
<http://localhost:8080/rest/test_versioning/0e/4b/a0/88/0e4ba088-ea77-4760-99ab-d57ead90bb6e>; rel="timegate",
<http://localhost:8080/rest/test_versioning/0e/4b/a0/88/0e4ba088-ea77-4760-99ab-d57ead90bb6e/fcr:versions/20180427204227>; rel="memento"; datetime="Fri, 27 Apr 2018 20:42:27 GMT",
<http://localhost:8080/rest/test_versioning/0e/4b/a0/88/0e4ba088-ea77-4760-99ab-d57ead90bb6e/fcr:versions>; rel="self"; type="application/link-format",

[/sw/var/www/fcrepo4/fedora4-tests] 
> curl -i -XPOST -ufedoraAdmin:fedoraAdmin '-HContent-type: image/jpeg' '-HMemento-Datetime: Tue, 23 May 2000 17:34:45 +0000' --data-binary @/sw/var/www/fcrepo4/fedora4-tests/resources/basic_image.jpg http://localhost:8080/rest/test_versioning/0e/4b/a0/88/0e4ba088-ea77-4760-99ab-d57ead90bb6e/fcr:versions
HTTP/1.1 100 Continue

HTTP/1.1 409 Conflict
Date: Fri, 27 Apr 2018 20:44:07 GMT
Set-Cookie: JSESSIONID=1p5a0usmeapelvqms1vtr4lnt;Path=/
Expires: Thu, 01 Jan 1970 00:00:00 GMT
Set-Cookie: rememberMe=deleteMe; Path=/; Max-Age=0; Expires=Thu, 26-Apr-2018 20:44:07 GMT
Content-Type: text/plain;charset=utf-8
Content-Length: 45
Server: Jetty(9.3.1.v20150714)

Memento with provided datetime already exists
```

**Comment by bbpennel** [comment]
@whikloj @awoods  looks like the reason why you're getting a conflict when trying to recreate the memento is because a tombstone gets left behind. I'm not sure there's much point to keeping a tombstone for a memento, so should we make it so a delete to a memento skips the tombstone?

**Comment by whikloj** [comment]
I'm a +1 to not having a tombstone for a memento.

**Comment by whikloj** [comment]
@bbpennel I have two issues.
1) When you delete a binary Memento, you seem to be losing the corresponding Description memento. I'm not sure that is correct behaviour.
2) When trying to `PUT` over-top of a binary Memento we are getting a 415 instead of 405.
```
curl --no-keepalive -i -XPATCH -ufedoraAdmin:fedoraAdmin -H"Content-type: image/jpeg" --data-binary "@/sw/var/www/fcrepo4/fedora4-tests/resources/basic_image.jpg" http://localhost:8080/rest/test_versioning/c9/a3/ce/82/c9a3ce82-5c67-439a-aa76-40fb65a9a7cf/fcr:versions/20180430173148
HTTP/1.1 100 Continue

HTTP/1.1 415 Unsupported Media Type
Date: Mon, 30 Apr 2018 17:32:00 GMT
Set-Cookie: JSESSIONID=111p7weilupt9rdb9b66zjetn;Path=/
Expires: Thu, 01 Jan 1970 00:00:00 GMT
Set-Cookie: rememberMe=deleteMe; Path=/; Max-Age=0; Expires=Sun, 29-Apr-2018 17:32:00 GMT
Content-Type: text/plain;charset=utf-8
Content-Length: 31
Server: Jetty(9.3.1.v20150714)

HTTP 415 Unsupported Media Type
```

File: fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/FedoraBinaryImpl.java (hunk 1 above)
Comment (awoods) [reply fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/FedoraBinaryImpl.java]: At a minimum, a detailed comment explaining the structure that requires `../../` would help in future maintenance.

Comment (awoods) [reply fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/FedoraBinaryImpl.java]: I see you have taken a different approach here:
https://github.com/fcrepo4/fcrepo4/pull/1344/files#diff-7aee50d5030524028026aa94cd75f350R419

File: fcrepo-http-api/src/main/java/org/fcrepo/http/api/FedoraVersioning.java [hunk 5]
```diff
@@ -185,29 +186,40 @@ public Response addVersion(@HeaderParam(MEMENTO_DATETIME_HEADER) final String da
...
+                        memento = versionService.createBinaryVersion(session.getFedoraSession(),
+                                binaryResource, mementoInstant, storagePolicyDecisionPoint);
+                    } else {
+                        memento = createBinaryMementoFromRequest(binaryResource, mementoInstant,
+                                requestBodyStream, digest);
+                    }
+                }
+                // Create rdf memento if the request resource was an rdf resource or a binary from the
+                // current version of the original resource.
+                if (!isBinary || (isBinary && createFromExisting)) {
```
Comment (awoods) [inline fcrepo-http-api/src/main/java/org/fcrepo/http/api/FedoraVersioning.java]: Can remove `isBinary` from `(isBinary && createFromExisting)`... since it will always be `true` given the failure of the first condition.

File: fcrepo-http-api/src/test/java/org/fcrepo/integration/http/api/FedoraVersioningIT.java [hunk 6]
```diff
@@ -118,7 +116,7 @@
 public class FedoraVersioningIT extends AbstractResourceIT {
 
     public static final DateTimeFormatter MEMENTO_DATETIME_ID_FORMATTER =
-        DateTimeFormatter.ofPattern("yyyyMMddHHmmss").withZone(ZoneId.of("GMT"));
+            DateTimeFormatter.ofPattern("yyyyMMddHHmmss").withZone(ZoneId.of("GMT"));
```
Comment (awoods) [inline fcrepo-http-api/src/test/java/org/fcrepo/integration/http/api/FedoraVersioningIT.java]: Quite a bit of unrelated formatting changes in this file. It would make the review (and merge conflicts) easier without such updates.

File: fcrepo-http-api/src/test/java/org/fcrepo/integration/http/api/FedoraVersioningIT.java [hunk 7]
```diff
@@ -597,22 +648,17 @@ public void testCreateVersionOfBinary() throws Exception {
...
-    public void testCreateVersionOfBinaryWithDatetime() throws Exception {
-        createVersionedBinary(id);
-
-        final HttpPost createVersionMethod = postObjMethod(id + "/" + FCR_VERSIONS);
-        createVersionMethod.addHeader(MEMENTO_DATETIME_HEADER, MEMENTO_DATETIME);
+        // Verifying that the associated description memento was created
+        final String descriptionMementoUri = mementoUri.replace("fcr:versions", "fcr:metadata/fcr:versions");
 
-        try (final CloseableHttpResponse response = execute(createVersionMethod)) {
-            assertEquals("Must reject memento creation without Content Type",
```
Comment (awoods) [inline fcrepo-http-api/src/test/java/org/fcrepo/integration/http/api/FedoraVersioningIT.java]: This is still a useful test to verify, no? i.e. requirement for a content-type when creating a memento from a provided body.

File: fcrepo-http-api/src/test/java/org/fcrepo/integration/http/api/FedoraVersioningIT.java [hunk 8]
```diff
@@ -648,40 +694,179 @@ public void testCreateVersionOfBinaryWithBody() throws Exception {
...
+        try (final CloseableDataset dataset = getDataset(new HttpGet(mementoUri))) {
+            final DatasetGraph results = dataset.asDatasetGraph();
+
+            final Node mementoSubject = createURI(subjectUri);
+
+            assertFalse("Property added to original must not appear in memento",
+                    results.contains(ANY, mementoSubject, DC.title.asNode(), ANY));
+            assertFalse("Memento type should not be visible",
+                    results.contains(ANY, mementoSubject, RDF.type.asNode(), MEMENTO_TYPE_NODE));
+            assertFalse("Property added to original must not appear in memento",
```
Comment (awoods) [inline fcrepo-http-api/src/test/java/org/fcrepo/integration/http/api/FedoraVersioningIT.java]: This test appears to be a duplicate of line 733 above.

File: fcrepo-http-api/src/test/java/org/fcrepo/integration/http/api/FedoraVersioningIT.java [hunk 9]
```diff
@@ -648,40 +694,179 @@ public void testCreateVersionOfBinaryWithBody() throws Exception {
...
+        assertMementoUri(mementoUri, descriptionUri);
+
+        setDescriptionProperty(id, null, DC.title.getURI(), "Updated");
+
+        try (final CloseableDataset dataset = getDataset(new HttpGet(mementoUri))) {
+            final DatasetGraph results = dataset.asDatasetGraph();
+
+            final Node mementoSubject = createURI(subjectUri);
+
+            assertFalse("Property added to original must not appear in memento",
```
Comment (awoods) [inline fcrepo-http-api/src/test/java/org/fcrepo/integration/http/api/FedoraVersioningIT.java]: It would be helpful to have at least one `assertTrue` test to ensure that some expected triples are present.

File: fcrepo-http-api/src/test/java/org/fcrepo/integration/http/api/FedoraVersioningIT.java [hunk 10]
```diff
@@ -648,40 +694,179 @@ public void testCreateVersionOfBinaryWithBody() throws Exception {
...
+        final String mementoUri = createContainerMementoWithBody(descriptionUri, MEMENTO_DATETIME);
+        assertMementoUri(mementoUri, descriptionUri);
+
+        try (final CloseableDataset dataset = getDataset(new HttpGet(mementoUri))) {
+            final DatasetGraph results = dataset.asDatasetGraph();
 
-        createObject(pid);
+            final Node mementoSubject = createURI(mementoUri);
+
+            assertFalse("Memento type should not be visible",
```
Comment (awoods) [inline fcrepo-http-api/src/test/java/org/fcrepo/integration/http/api/FedoraVersioningIT.java]: Is there also an `assertTrue` test that we can include that proves that there is a property that we expect to find?

File: fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/services/VersionService.java [hunk 11]
```diff
@@ -69,17 +72,34 @@ FedoraResource createVersion(FedoraSession session, FedoraResource resource,
...
+    FedoraBinary createBinaryVersion(FedoraSession session,
+            FedoraBinary resource,
+            Instant dateTime,
+            InputStream contentStream,
+            Collection<URI> checksums,
+            StoragePolicyDecisionPoint storagePolicyDecisionPoint)
+            throws InvalidChecksumException;
 
+    /**
+     * Explicitly creates a version of a binary resource. If no contentStream is provided, then
```
Comment (awoods) [inline fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/services/VersionService.java]: Comment seems unfinished.

File: fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/services/VersionService.java [hunk 12]
```diff
@@ -69,17 +72,34 @@ FedoraResource createVersion(FedoraSession session, FedoraResource resource,
      * Explicitly creates a version of a binary resource. If no contentStream is provided, then
```
Comment (awoods) [inline fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/services/VersionService.java]: Comment seems unfinished.

File: fcrepo-http-api/src/main/java/org/fcrepo/http/api/FedoraVersioning.java (hunk 3 above)
Comment (whikloj) [reply fcrepo-http-api/src/main/java/org/fcrepo/http/api/FedoraVersioning.java]: Sorry the question was around the fact that on line 193 you check if `resource instanceof FedoraBinary` and if it is, then you cast it as a `FedoraBinary`.

File: fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/FedoraBinaryImpl.java [hunk 13]
```diff
@@ -441,15 +469,21 @@ private static void decorateContentNode(final Node dsNode, final Node descNode,
...
         }
     }
 
     private boolean hasDescriptionProperty(final String relPath) {
         try {
+            final Node descNode = getDescriptionNodeOrNull();
+            if (descNode == null) {
+                return false;
+            }
             return getDescriptionNode().hasProperty(relPath);
```
Comment (whikloj) [inline fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/FedoraBinaryImpl.java]: Probably could just be `return descNode.hasProperty(relPath);` now.

File: fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/services/VersionServiceImpl.java [hunk 14]
```diff
@@ -206,28 +228,67 @@ private Triple convertToInternalReference(final Triple t,
...
+    @Override
+    public FedoraBinary createBinaryVersion(final FedoraSession session,
+            final FedoraBinary resource,
             final Instant dateTime,
             final InputStream contentStream,
-            final String filename,
-            final String mimetype,
-            final Collection<URI> checksums) {
+            final Collection<URI> checksums,
+            final StoragePolicyDecisionPoint storagePolicyDecisionPoint) throws InvalidChecksumException {
```
Comment (awoods) [inline fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/services/VersionServiceImpl.java]: I am not seeing where `storagePolicyDecisionPoint` is used. Is there any reason not to remove it from the interface?

File: fcrepo-http-api/src/test/java/org/fcrepo/integration/http/api/FedoraVersioningIT.java (hunk 7 above)
Comment (bbpennel) [reply fcrepo-http-api/src/test/java/org/fcrepo/integration/http/api/FedoraVersioningIT.java]: While working on this branch I realized that there didn't appear to be a reason to provide a content type when submitting a historic binary memento. The description is where the content type and other file metadata is provided, which is submitted in a second request, so the file info provided when creating the binary memento would just be ignored (except for checksums which are verified). For that reason I removed the requirement that a content type be provided for binaries.

File: fcrepo-http-api/src/test/java/org/fcrepo/integration/http/api/FedoraVersioningIT.java (hunk 9 above)
Comment (bbpennel) [reply fcrepo-http-api/src/test/java/org/fcrepo/integration/http/api/FedoraVersioningIT.java]: Ah I think i had that in one of my three branches... will add

File: fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/services/VersionServiceImpl.java (hunk 14 above)
Comment (bbpennel) [reply fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/services/VersionServiceImpl.java]: I don't actually know what its used for, but that was unintentional, it's being passed along to setContent now.

File: fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/FedoraBinaryImpl.java (hunk 1 above)
Comment (bbpennel) [reply fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/FedoraBinaryImpl.java]: It was a shorter path :) If its preferable I can switch this to 
```
return node.getParent().getParent().getNode(FEDORA_DESCRIPTION)
                        .getNode(LDPCV_TIME_MAP).getNode(mementoName);
```
And over in `NonRdfSourceDescriptionImpl.getContentNode` to
```
return node.getParent().getParent().getParent().getNode(LDPCV_TIME_MAP).getNode(mementoName);
```
(feel sorry for anyone that ends up unit testing that)

File: fcrepo-http-api/src/main/java/org/fcrepo/http/api/FedoraVersioning.java (hunk 3 above)
Comment (bbpennel) [reply fcrepo-http-api/src/main/java/org/fcrepo/http/api/FedoraVersioning.java]: `resource` is declared as a `FedoraResource`, which FedoraBinary extends, so I'm just narrowing it down to a more specific implementation.

File: fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/FedoraBinaryImpl.java (hunk 1 above)
Comment (bbpennel) [reply fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/FedoraBinaryImpl.java]: Actually, decided to save us another round of review buttons and switched https://github.com/fcrepo4/fcrepo4/pull/1344/commits/2fe2a0e9048afe6df4ae756abb1e220995bce4f7

File: fcrepo-http-api/src/test/java/org/fcrepo/integration/http/api/FedoraVersioningIT.java [hunk 15]
```diff
@@ -399,6 +396,36 @@ public void testCreateVersionDuplicateMementoDatetime() throws Exception {
...
+        assertEquals("Delete memento must be removed",
+                NOT_FOUND.getStatusCode(), getStatus(new HttpGet(mementoUri)));
+
+        final String recreatedUri = createContainerMementoWithBody(subjectUri, MEMENTO_DATETIME);
+        assertEquals("Recreated memento must exist",
+                OK.getStatusCode(), getStatus(new HttpGet(recreatedUri)));
+    }
+
+    @Test
+    public void testDeleteAndPostBinaryMemento() throws Exception {
```
Comment (awoods) [inline fcrepo-http-api/src/test/java/org/fcrepo/integration/http/api/FedoraVersioningIT.java]: Would you mind either adding content to these empty tests, creating JIRAs for them, or removing them?
